API / Improve associated resources search

fxprunayre opened the pull request:
### Current situation
Current API operation is: GET /api/records/{uuid}/related

The operation extract relations for records. We mix here 2 types of items
* Metadata records
* Online source / overview / additional document (eg. legend)

For metadata records, we have 3 types of links
* associated resource in the XML eg. parent
* associated resource which requires a search to find them eg. children
* associated resource pointing to remote records

Processing is
* Get XML 
* Collect record links from XML
* Collect record links using searches
* Append remote records links
* Check if records are in current portal or not
* Build a Java XML response processed by XSLT relation.xsl 

### Goals

* Related API returns online links and overviews which are already in the index document (cf. `link` and `overview` properties) - use the index document for links and overviews and use the new API for associated records 
* A metadata record in the related API has a different model than in a search response - Associated API returns now index document for easier use on client side
* Improve speed - reduce HTTP calls, rework search for related docs

### Proposal

Add a GET /api/records/{uuid}/associated operation which returns only records (no more links already available through the search API)
* The related records are the document in the index with additional properties about the type of link
* The search is faster by searching all references in the record, retrieving documents based on user privileges and then applying additional properties (eg. portal filter, remote records). Removal of XSL processing.

On client side, the response can be treated like other `Metadata` objects which makes search response and this API consistent. It also remove duplicated ways of retrieving links and overview.

Deprecate GET /api/records/{uuid}/related

Editor associated panel is also using associated API for related records:

![image](https://user-images.githubusercontent.com/1701393/165296517-5754b8bc-5835-4517-8dff-57b1ddf35e81.png)


### UI changes & improvements 

* More like this / Exclude associated records from the more like this results - they are already displayed in the record view page

* Add warning message while linking a dataset (that current user can't edit) to a service

![image](https://user-images.githubusercontent.com/1701393/165295429-f494c773-ab32-471b-a71a-04a600e0a47e.png)

* Avoid duplicates when relation are created in both direction (using associatedResources). In those cases records may appear in both children/parent/siblings and in associated.

* Display revision of current record (encoded using revisionOf relation)

![image](https://user-images.githubusercontent.com/1701393/165891729-f5485af2-35e1-4cfc-8dce-2e392df37157.png)

* Children can be encoded using isComposedOf relation


### Performances

The associated API is faster than the related API (the response is a bigger as it contains the index document with more information). 

```
Concurrency Level:      1
Complete requests:      20
Failed requests:        0


Document Path:          /geonetwork/srv/api/records/5b165b25-0b59-4c64-8908-3a82ae1ba304/related
Document Length:        69383 bytes
Time taken for tests:   3.267 seconds
Requests per second:    6.12 [#/sec] (mean)
Time per request:       163.352 [ms] (mean)

Document Path:          /geonetwork/srv/api/records/5b165b25-0b59-4c64-8908-3a82ae1ba304/associated
Document Length:        248303 bytes
Time taken for tests:   0.828 seconds
Requests per second:    24.16 [#/sec] (mean)
Time per request:       41.388 [ms] (mean)

```

The associated resources can also be added directly to a search response with http://localhost:8080/geonetwork/srv/api/search/records/_search?relatedType=parent&relatedType=children&relatedType=sources&relatedType=hassources&relatedType=brothersAndSisters&relatedType=services&relatedType=datasets&relatedType=siblings&relatedType=associated&relatedType=fcats&relatedType=related

![image](https://user-images.githubusercontent.com/1701393/165296332-8921fb1b-d99e-4fa2-8033-beefff8f91f9.png)



Related records can be directly displayed in search (instead of running another call to the Related API
![image](https://user-images.githubusercontent.com/1701393/165233044-71016eb5-af03-46b3-a29a-b61117f9e3e1.png)



### Future related work

* Links are not indexed using multilingual structure https://github.com/geonetwork/core-geonetwork/blob/main/schemas/iso19139/src/main/plugin/iso19139/index-fields/index.xsl#L1022-L1024

MichelGabriel commented:
Looks great, I have just a little comment about this part:

> Related records can be directly displayed in search (instead of running another call to the Related API

Could there be a small title above the associated links so the user knows what these links are?

josegar74 commented on schemas/iso19115-3.2018/src/main/plugin/iso19115-3.2018/process/collection-updater.xsl:
```diff
@@ -32,7 +32,7 @@
              select="''"/>
 
   <!-- (optional) The association type. Default: crossReference. -->
```
Update comment with the changed default value `isComposedOf`

josegar74 commented on schemas/iso19139/src/main/plugin/iso19139/process/collection-updater.xsl:
```diff
@@ -28,7 +28,7 @@
              select="''"/>
 
   <!-- (optional) The association type. Default: crossReference. -->
```
Update comment with the changed default value `isComposedOf`

josegar74 commented on services/src/main/java/org/fao/geonet/api/records/MetadataUtils.java:
```diff
@@ -99,17 +90,312 @@ public class MetadataUtils {
-    private static final String ORIGIN_REMOTE = "remote";
+    public static class RelatedTypeDetails {
+        private String query;
+        private Set<String> expectedRecords = new HashSet<>();
```
What are the expected records? Records defined in the metadata relations?

fxprunayre replied:
While working on this we were also discussing a way to track bad relations (eg. if a record is removed but is still linked) so we may have deadlinks at some point. The search is built in most case on top of a list of UUIDs to search for and if user is admin expectedRecords is supposed to match the number of UUID to found. We did not complete that part as the link analysis partially analyse that aspect too (but also have drawbacks eg. private records can not be checked). So further work is required in that area.

josegar74 commented on services/src/test/java/org/fao/geonet/api/records/MetadataApiTest.java:
```diff
@@ -22,6 +22,7 @@
  */
 package org.fao.geonet.api.records;
 
+import static org.apache.fop.fonts.type1.AdobeStandardEncoding.p;
```
Is this required?

josegar74 commented:
Some issues found:

1) With the parent / child relation:

- The child metadata page displays the link to the parent.
- The parent metadata page *does not* display the link to the child.

2) When editing a metadata I see the following failed request, it seems that the request to retrieve thumbnails is using the new API and it shouldn't be the case:

http://localhost:8080/geonetwork/srv/api/records/b64ca2fb-fe94-4fc7-8d07-503846cb1976/associated?type=thumbnail

```
{"message":"MethodArgumentTypeMismatchException","code":"runtime_exception","description":"Failed to convert value of type 'java.lang.String' to required type 'org.fao.geonet.api.records.model.related.RelatedItemType[]'; nested exception is org.springframework.core.convert.ConversionFailedException: Failed to convert from type [java.lang.String] to type [org.fao.geonet.api.records.model.related.RelatedItemType] for value 'thumbnail'; nested exception is java.lang.IllegalArgumentException: No enum constant org.fao.geonet.api.records.model.related.RelatedItemType.thumbnail"}
```

3) In the editor I added an association to a remote record, using the Geonetwork url for the metadata landing page, but the relation is not created. No error is returned or added to the log file.

fxprunayre commented:
> Could there be a small title above the associated links so the user knows what these links are?

We used the same icons as in other places (record view and editor). Not sure an extra label is necessary but it could be something like 

![image](https://user-images.githubusercontent.com/1701393/166232324-e202013e-1be4-4a2a-b387-0eef84cc178c.png)

@MichelGabriel ? 


> When editing a metadata I see the following failed request, it seems that the request to retrieve thumbnails is using the new API and it shouldn't be the case:

Indeed for ISO19139 in the overview manager directive. Fixed

> With the parent / child relation:
> The child metadata page displays the link to the parent.
> The parent metadata page does not display the link to the child.

If the parent is a serie it should be fine. I suppose your test was linking a dataset to another dataset ? So it relates more to the work we did in https://github.com/titellus/core-geonetwork/pull/81. Adding children for dataset above associated records.


> In the editor I added an association to a remote record, using the Geonetwork url for the metadata landing page, but the relation is not created. No error is returned or added to the log file.

Which type of GeoNetwork URL did you used ? Can you check the XML, if no error, the relation was probably added but something is wrong in the encoding ?

MichelGabriel commented:
Thanks, I like it (it's more for 'don't let me think')
